File: src/providers/provider.ts
import type { NormalizedError, NormalizedRow, ProviderId } from "../types.js";

export type AcquireContext = {
  refresh: boolean;
  debug: boolean;
};

export type ProviderResult = {
  rows: NormalizedRow[];
  errors: NormalizedError[];
  debug?: Record<string, unknown>;
};

export interface Provider {
  id: ProviderId;
  acquire(ctx: AcquireContext): Promise<ProviderResult>;
}

File: crates/qmeter-providers/src/provider.rs
use qmeter_core::types::{NormalizedError, NormalizedRow, ProviderId};
use serde_json::Value;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct AcquireContext {
    pub refresh: bool,
    pub debug: bool,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ProviderResult {
    pub rows: Vec<NormalizedRow>,
    pub errors: Vec<NormalizedError>,
    pub debug: Option<Value>,
}

pub trait Provider {
    fn id(&self) -> ProviderId;
    fn acquire(&self, ctx: AcquireContext) -> ProviderResult;
}

